Minda SV Annotation
===================

This folder contain a set of scripts that annotates SV calls ensemble
generated by Minda with various challenging scenarios. We currently annotate the following categories.
More detailed description is availabe in the 
[Severus manuscript](https://www.medrxiv.org/content/10.1101/2024.03.22.24304756v1)

* Insertions
* SVs with breakpoints at matching homologous repeats
* SVs inside segmental duplications
* SVs inside VNTRs	
* SVs with low allelic frequency
* Short SVs (<100bp)
* Duplicated SVs
* SV chains

To run annotation, you'll need Minda output directory and a directory
with annotation for your reference genome. Links to download existing
annotations are below:

```
grch38  XXX
```

For example, if you are using grch38:

```
wget XXX.tar.gz
tar -xvf XXX.tar.gz
./minda_stratify.py XXX minda_support.tsv 11
```

where `minda_support.tsv` is a support file generated by Minda, and `11` is the number of callsets
used to create the Minda ensemble.

Currently, genome annotations include the following:
* Chromosome lengths (produces using `samtools faidx`)
* Common repeat annotations (produces using `RepeatMasker`)
* Segmental duplications annotation 
* VNTR annotations (produced using [findTandemRepeats](https://github.com/PacificBiosciences/pbsv/tree/master/annotations)
